plugins/woocommerce-admin/client/core-profiler/index.tsx

```diff
@@ -616,32 +672,206 @@ export const coreProfilerStateMachineDefinition = createMachine( {
 			} ),
 			meta: {
 				progress: 70,
 			},
 		},
-		extensions: {
+		pluginsSkipped: {
+			entry: assign( {
+				loader: {
+					progress: 80,
+				},
+			} ),
+			invoke: {
+				src: () => {
+					dispatch( ONBOARDING_STORE_NAME ).updateProfileItems( {
+						plugins_page_skipped: true,
+						completed: true,
+					} );
+					return promiseDelay( 3000 );
+				},
+				onDone: {
+					actions: [ 'redirectToWooHome' ],
+				},
+			},
+			meta: {
+				component: Loader,
+			},
+		},
+		plugins: {
+			entry: [ 'recordTracksPluginsViewed' ],
 			on: {
-				EXTENSIONS_COMPLETED: {
-					target: 'settingUpStore',
+				PLUGINS_PAGE_SKIPPED: {
+					actions: [ 'recordTracksPluginsSkipped' ],
+					target: 'pluginsSkipped',
+				},
+				PLUGINS_INSTALLATION_REQUESTED: {
+					target: 'installPlugins',
+					actions: [
+						assign( {
+							pluginsSelected: (
+								_context,
+								event: PluginsInstallationRequestedEvent
+							) => event.payload.plugins,
+						} ),
+					],
 				},
 			},
 			meta: {
 				progress: 80,
-				component: Extensions,
+				component: Plugins,
+			},
+		},
+		postPluginInstallation: {
+			invoke: {
+				src: async ( _context, event ) => {
+					return await dispatch(
+						ONBOARDING_STORE_NAME
+					).updateProfileItems( {
+						business_extensions: event.payload.installationCompletedResult.installedPlugins.map(
+							( extension: InstalledPlugin ) => extension.plugin
+						),
+						completed: true,
+					} );
+				},
+				onDone: {
+					actions: 'redirectToWooHome',
+				},
+			},
+			meta: {
+				component: Loader,
+				progress: 100,
+			},
+		},
+		installPlugins: {
+			on: {
+				PLUGIN_INSTALLED_AND_ACTIVATED: {
+					actions: [
+						assign( {
+							loader: (
+								_context,
+								event: PluginInstalledAndActivatedEvent
+							) => {
+								const progress = Math.round(
+									( event.payload.installedPluginIndex /
+										event.payload.pluginsCount ) *
+										100
+								);
+
+								let stageIndex = 0;
+
+								if ( progress > 30 ) {
+									stageIndex = 1;
+								} else if ( progress > 60 ) {
+									stageIndex = 2;
+								}
+
+								return {
+									useStages: 'plugins',
+									progress,
+									stageIndex,
+								};
+							},
+						} ),
+					],
+				},
+				PLUGINS_INSTALLATION_COMPLETED_WITH_ERRORS: {
+					target: 'prePlugins',
+					actions: [
+						assign( {
+							pluginsInstallationErrors: ( _context, event ) =>
+								event.payload.errors,
+						} ),
+						( _context, event ) => {
+							recordEvent(
+								'storeprofiler_store_extensions_installed_and_activated',
+								{
+									success: false,
+									failed_extensions: event.payload.errors.map(
+										( error: PluginInstallError ) =>
+											getPluginTrackKey( error.plugin )
+									),
+								}
+							);
+						},
+					],
+				},
+				PLUGINS_INSTALLATION_COMPLETED: {
+					target: 'postPluginInstallation',
+					actions: [
+						( _context, event ) => {
```

Comment:
I think we should consider refactoring this by splitting actions into separate functions and putting them in [`coreProfilerMachineActions`](https://github.com/woocommerce/woocommerce/blob/bd2b34bc7a1d6ec901a04c9c78b6ae74bdfbb77a/plugins/woocommerce-admin/client/core-profiler/index.tsx#L378). But it's fine for now.

Reply:
Thanks! Yeah, let's discuss how we want to extract these actions and services later. We probably need guidance from @rjchow as well.

Reply:
No specific guidance for this! Just make it a function and reference it by label. I'm planning on reviewing the sprawl during cooldown next week, so will take care of that then